### PR TITLE
Add support for sql as a header to create or replace

### PR DESCRIPTION
**Pull request by kconvey**
#1879

Any SQL that does not work within `create or replace table` can be extracted and injected ahead of `create or replace table` by using a call block like:
```
{{%- call sql_header(config) -%}}
CREATE TEMPORARY FUNCTION yes_no_to_boolean(answer STRING)
RETURNS BOOLEAN AS (
  CASE
  WHEN LOWER(answer) = 'yes' THEN True
  WHEN LOWER(answer) = 'no' THEN False
  ELSE NULL
  END
);
{{%- endcall -%}}
```

The resulting SQL will look like:
```
CREATE TEMPORARY FUNCTION yes_no_to_boolean(answer STRING)
RETURNS BOOLEAN AS (
  CASE
  WHEN LOWER(answer) = 'yes' THEN True
  WHEN LOWER(answer) = 'no' THEN False
  ELSE NULL
  END
);

create or replace table `project.dataset.table`
...
```

This is most obviously useful for BigQuery's temporary UDFs, which do not have a perfect dbt analogue. dbt's support for persistent UDFs or use of macros can likely stand-in in most cases, but this feature facilitates dbt execution of SQL that was written (or generated) to BQ's SQL standard, rather than Jinja-templated SQL.

### CODE REVIEW

**Comment by kconvey** [comment]
@drewbanin Think I got to all of your suggestions. Let me know if 'common.sql' is a reasonable place to put the `set_sql_header()` macro across adapters. 

Attempted to add an integration test for bigquery that creates an innocuous table as a header with a `create or replace table`, that should fail if it is not extracted properly (`create or replace table ( create or repalce table() )` vs `create or replace table() create or replace table()`. I think a similar pattern should work with the other adapters, but don't know enough about how they work / how the integration tests are set up to best implement them.

**Comment by sdepablos** [comment]
Would this work for anything database specific, like [BigQuery SQL scripting](https://cloud.google.com/bigquery/docs/reference/standard-sql/scripting)?

File: test/integration/022_bigquery_test/models/sql_header_model.sql [hunk 1]
```diff
@@ -0,0 +1,8 @@
+{# This will fail if it is not extracted correctly #}
+{% call set_sql_header(config) %}
+  	create or replace table {{ ref('table_model') }} as (
```
Comment (drewbanin) [inline test/integration/022_bigquery_test/models/sql_header_model.sql]: can you just make this a simple temporary UDF which is referenced in the model body below? I think that will more clearly exercise this new feature

Comment (kconvey) [reply test/integration/022_bigquery_test/models/sql_header_model.sql]: Definitely more clear. Good suggestion!

**Comment by kconvey** [comment]
@drewbanin Can't believe I would forget Snowflake :/

 Better test + Snowflake should be added now.

@sdepablos Tried it out and it seems like this works for BigQuery SQL scripting! Originally only had temporary UDFs as a clear use case, but glad there are others.

The following should work, as an example:
```
{{ call set_sql_header(config) }}
  -- Declare a variable to hold names as an array.
  DECLARE top_names ARRAY<STRING>;
  -- Build an array of the top 100 names from the year 2017.
  SET top_names = (
    SELECT ARRAY_AGG(name ORDER BY number DESC LIMIT 100)
    FROM `bigquery-public-data`.usa_names.usa_1910_current
    WHERE year = 2017
  );
{{ endcall }}
```

File: test/integration/022_bigquery_test/models/sql_header_model.sql [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{# This will fail if it is not extracted correctly #}
+{% call set_sql_header(config) %}
+  	CREATE TEMPORARY FUNCTION a_to_b(str STRING)
+	RETURNS STRING AS (
+	  CASE
+	  WHEN LOWER(str) = 'a' THEN 'b'
+	  ELSE str
+	  END
+	)
```
Comment (drewbanin) [inline test/integration/022_bigquery_test/models/sql_header_model.sql]: I think you might just need a semicolon here - I see
```
Completed with 1 error and 0 warnings:

Database Error in model sql_header_model (models/sql_header_model.sql)
  Syntax error: Unexpected keyword CREATE at [12:3]
  compiled SQL at target/run/test/sql_header_model.sql
```

in the CI build

Comment (kconvey) [reply test/integration/022_bigquery_test/models/sql_header_model.sql]: Should be fixed!

**Comment by drewbanin** [comment]
Ok - thanks! I just kicked off the tests, but it does look like there's a small merge conflict introduced by the addition of BigQuery labels. Can you take a quick look at that?

**Comment by kconvey** [comment]
Pulled the latest changes and labeled the column in the integration test (so hopefully that passes). May need you to kick off tests again.

Thanks!

**Comment by drewbanin** [comment]
Meep! I'm seeing
```
Database Error in model sql_header_model (models/sql_header_model.sql)
  Creating logical views with temporary user-defined functions is not supported
  compiled SQL at target/run/test/sql_header_model.sql
```

in the integration tests. Should we configure that model to be materialized as a table instead of a view? I did't realize this was a limitation of temporary UDFs.

I think you can ignore the postgres error - that seems like a network blip to me

**Comment by kconvey** [comment]
Should be configured as a table now.

**Comment by drewbanin** [comment]
Gosh, lots of little tweaks to make around the test suite here. FYI, we plan on making PRs build for community contributed PRs to help tighten this CI loop. 

I think we need to bump the number of expected tests from 7 to 8 here. We do this to make sure that all of the specified models in a test actually run, and this PR adds a new model. Check out the code here:

https://github.com/fishtown-analytics/dbt/blob/6be4ee00aaf5e794005835d0144e4e6a3e87834e/test/integration/022_bigquery_test/test_simple_bigquery_view.py#L56-L69

And an error from the CI build here:
```
______________ TestUnderscoreBigQueryRun.test_bigquery_run_twice _______________
[gw2] linux -- Python 3.8.0 /home/dbt_test_user/project/.tox/integration-bigquery-py38/bin/python

self = <test_simple_bigquery_view.TestUnderscoreBigQueryRun testMethod=test_bigquery_run_twice>

    @use_profile('bigquery')
    def test_bigquery_run_twice(self):
        self.run_dbt(['seed'])
        results = self.run_dbt()
>       self.assertEqual(len(results), 7)
E       AssertionError: 8 != 7
```

We'll want to change each instance of `7` in the snippet above to an `8`.

**Comment by kconvey** [comment]
Should have remembered that one from the labels PR. Thanks for catching!

**Comment by drewbanin** [comment]
@beckjake can you try kicking off the tests here? Azure doesn't seem to want to run this with env vars for me :/

**Comment by beckjake** [comment]
@drewbanin I tried to kick it off manually - we'll see if it works, I guess.

**Comment by beckjake** [comment]
The tests are successfully running here: https://dev.azure.com/fishtown-analytics/dbt/_build/results?buildId=1260&view=results
I have no idea how to make them show up in github though 🤷‍♂

**Comment by zestyping** [comment]
For future visitors to this issue: I tried a few ways of defining temporary BigQuery functions for use in model definitions, and arrived at the following setup, which you might also enjoy.

1. Define the functions in a macro, and inside the macro call `set_sql_header` like this:
```
{% macro define_bigquery_functions() %}
{% call set_sql_header(config) %}

create temporary function int(value any type)
returns int64 as (
  cast(cast(value as decimal) as int64)
);
```

2. Define a shorthand macro that wraps `config` like this:

```
{% macro table() %}
{{ define_bigquery_functions() }}
{{ config(materialized='table', **kwargs) }}
{% endmacro %}
```

3. Now you can just invoke `table()` instead of `config(materialized='table')` at the top of your model files, and you'll have access to your temporary functions!

Advantages of this method:
  - You don't have to specify the schema name in front of every function call
  - Actually reduces boilerplate instead of adding more of it
  - You'll never again wonder whether the parameter is supposed to be `materialize=` or `materialized=`

Disadvantages of this method:
  - You can't call these functions inside view definitions (BigQuery doesn't support temporary functions in views)

**Pull request merged**